apps/next-evans/lib/db/friends.ts

```diff
@@ -1,37 +1,60 @@
 import invariant from 'tiny-invariant';
+import type { Lang } from '@friends-library/types';
 import type { CustomCode } from './custom-code';
-import type { DocumentWithFriendMeta, FriendProps } from '../types';
+import type { DocumentWithMeta, FriendType } from '../types';
 import { LANG } from '../env';
+import { documentRegion, getPrimaryResidence } from '../residences';
 import { prisma } from './prisma';
 import getAllCustomCode from './custom-code';
 
-let friendsPromise: Promise<Record<string, FriendProps>> | null = null;
+let friendsPromise_en: Promise<Record<string, FriendType>> | null = null;
+let friendsPromise_es: Promise<Record<string, FriendType>> | null = null;
 
-export async function getAllFriends(): Promise<Record<string, FriendProps>> {
-  if (friendsPromise) {
-    return friendsPromise;
+export async function getAllFriends(
+  lang: Lang = LANG,
+): Promise<Record<string, FriendType>> {
+  if (lang === `en`) {
+    if (friendsPromise_en) {
+      return friendsPromise_en;
+    }
+    friendsPromise_en = Promise.all([getFriendsFromDB(lang), getAllCustomCode()]).then(
+      ([friends, customCode]) => addCustomCodeToFriends(friends, customCode),
+    );
+    return friendsPromise_en;
   }
-  friendsPromise = Promise.all([_getFriends(), getAllCustomCode()]).then(
+  if (friendsPromise_es) {
+    return friendsPromise_es;
+  }
+  friendsPromise_es = Promise.all([getFriendsFromDB(lang), getAllCustomCode()]).then(
     ([friends, customCode]) => addCustomCodeToFriends(friends, customCode),
   );
-  return friendsPromise;
+  return friendsPromise_es;
 }
 
 export default async function getFriend(
   friendSlug: string,
-): Promise<FriendProps | undefined> {
+): Promise<FriendType | undefined> {
   const friends = await getAllFriends();
   return friends[friendSlug];
 }
 
 // { `friendSlug/documentSlug`: Document }
-export async function getAllDocuments(): Promise<Record<string, DocumentWithFriendMeta>> {
-  const friends = await getAllFriends();
-  const documents: Record<string, DocumentWithFriendMeta> = {};
+export async function getAllDocuments(
+  lang: Lang = LANG,
+): Promise<Record<string, DocumentWithMeta>> {
+  const friends = await getAllFriends(lang);
+  const documents: Record<string, DocumentWithMeta> = {};
   Object.values(friends).forEach((friend) => {
     friend.documents.forEach((doc) => {
+      const primaryResidence = getPrimaryResidence(friend.residences);
+      const firstStay = primaryResidence?.durations[0];
+      const publicationDate = firstStay ? firstStay.start ?? firstStay.end : null;
       documents[`${friend.slug}/${doc.slug}`] = {
         ...doc,
+        publishedRegion: primaryResidence?.region
+          ? documentRegion(primaryResidence?.region)
+          : `Other`,
+        publishedDate: publicationDate ?? friend.died ?? friend.born ?? 1650,
```

Comment:
hmm... all this logic about publication date is kind of tricky and buried in the middle of this big `forEach`. i think i'd like to see this pulled out into a well-named function that could be understood in isolation and possibly later tested.

where exactly did you pull this logic from? were there tests on it in the old version?

Reply:
I think it's all coming from the api, so I don't know how to replicate it exactly; this is my best guess at how it works